asreviewcontrib/insights/algorithms.py

```diff
@@ -21,11 +20,34 @@ def _recall_values(labels, x_absolute=False, y_absolute=False):
 
 
 def _loss_value(labels):
-    positive_doc_ratio = sum(labels) / len(labels)
-    triangle_before_perfect_recall = positive_doc_ratio * 0.5
-    aera_under_recall_curve = metrics.auc(*_recall_values(labels))
-
-    return 1 - (triangle_before_perfect_recall + aera_under_recall_curve)
+    Ny = sum(labels)
+    Nx = len(labels)
+
+    # The best AUC represents the entire area under the perfect curve, which is
+    # the total area Nx * Ny, minus the area above the perfect curve (which is
+    # the sum of a series with a formula (Ny * Ny) / 2) plus 0.5 to account for
+    # the boundary.
+    best_auc = Nx * Ny - (((Ny * Ny) / 2) + 0.5)
+
+    # Compute recall values (y) based on the provided labels. We don't need x
+    # values because the points are uniformly spaced.
+    y = np.array(_recall_values(labels, x_absolute=True, y_absolute=True)[1])
+
+    # The actual AUC is calculated by approximating the area under the curve
+    # using the trapezoidal rule. (y[1:] + y[:-1]) / 2 takes the average height
+    # between consecutive y values, and we sum them up.
+    actual_auc = np.sum((y[1:] + y[:-1]) / 2)
+
+    # The worst AUC represents the area under the worst-case step curve, which
+    # is simply the area under the recall curve where all positive labels are
+    # clumped at the end, calculated as (Ny * Ny) / 2.
+    worst_auc = ((Ny * Ny) / 2)
```

Comment:
Similarly, this should be `worst_auc = (Ny * (Ny + 1)) / 2`

Reply:
Agreed